src/main/groovy/edu/oregonstate/mist/cardsapi/resources/CardsResource.groovy

```diff
@@ -65,4 +80,73 @@ class CardsResource extends Resource {
         }
 
     }
+
+    //Get cards by parameters
+    @GET
+    @Path ('')
+    @Produces(MediaType.APPLICATION_JSON)
+    Response getCards(@QueryParam("types") List<String> types,
+                      @QueryParam("name") Optional<String> name,
+                      @QueryParam("colors") List<String> colors,
+                      @QueryParam("rarities") List<String> rarities,
+                      @QueryParam("energyMin") Optional<Integer> energyMin,
+                      @QueryParam("energyMax") Optional<Integer> energyMax,
+                      @QueryParam("keywords") List<String> keywords,
+                      @QueryParam("number") Optional<Integer> number,
+                      @QueryParam("isRandom") Optional<Boolean> isRandom) {
+
+        if(!types) {
+            types = validTypes
+        } else {
+            List<String> invalidTypes = types - validTypes
+            if(invalidTypes) {
+                return badRequest("Invalid types: \'${invalidTypes.join("\', \'")}\'. " +
+                        "Valid types are: " +
+                        "skill, attack, power, status, curse").build()
+            }
+        }
+
+        if(!colors) {
+            colors = validColors
+        } else {
+            List<String> invalidColors = colors - validColors
+            if(invalidColors) {
+                return badRequest("Invalid colors: \'${invalidColors.join("\', \'")}\'. " +
+                        "Valid colors are: " +
+                        "red, green, blue, colorless.").build()
```

Comment:
Instead, you could reference the `validColors` list here so you don't have to repeat yourself. Could be something similar to `validColors.join(", ")` for example